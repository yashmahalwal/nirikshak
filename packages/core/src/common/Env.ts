/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
