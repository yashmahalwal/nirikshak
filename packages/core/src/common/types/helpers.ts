import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
