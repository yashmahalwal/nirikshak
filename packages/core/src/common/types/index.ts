export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
