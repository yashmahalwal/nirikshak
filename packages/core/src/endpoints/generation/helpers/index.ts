export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
