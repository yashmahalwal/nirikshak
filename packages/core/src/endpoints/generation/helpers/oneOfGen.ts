import faker from "faker";
import {
    Base,
    BodyType,
    WithModifiers,
    OneOfEntries,
} from "../../types/helpers";

// Return type with discriminated union
export type SelectedType<T extends Base | BodyType> = {
    type: T extends Base ? "base" : "body type";
    selection: T extends Base
        ? Base | WithModifiers<Base>
        : BodyType | WithModifiers<BodyType>;
};

// Function simply selects one of the types / fields
// and returns
export function generateOneOfEntries(
    input: OneOfEntries
): SelectedType<Base> | SelectedType<BodyType> {
    // Use faker.random.number instead of Math.random
    // So that randomness seed works

    // From the given selections, select an entry

    // If only types exists
    // Or we randomly choose type from choice b/w types and fields
    if (!("fields" in input))
        // Only type in the input
        return {
            type: "base",
            selection:
                input.types[
                    faker.random.number({
                        min: 0,
                        max: input.types.length - 1,
                        precision: 1,
                    })
                ],
        };

    // If only fields exists
    if (!("types" in input))
        return {
            type: "body type",
            selection:
                input.fields[
                    faker.random.number({
                        min: 0,
                        max: input.fields.length - 1,
                        precision: 1,
                    })
                ],
        };

    // Select either type or field
    return faker.random.boolean()
        ? {
              type: "body type",
              selection:
                  input.fields[
                      faker.random.number({
                          min: 0,
                          max: input.fields.length - 1,
                          precision: 1,
                      })
                  ],
          }
        : {
              type: "base",
              selection:
                  input.types[
                      faker.random.number({
                          min: 0,
                          max: input.types.length - 1,
                          precision: 1,
                      })
                  ],
          };
}
