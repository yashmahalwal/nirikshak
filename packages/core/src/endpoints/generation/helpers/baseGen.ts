import { Base } from "../../types/helpers";
import { ResourceInstance } from "../../../resource/types/helper";
import { SchemaHelpers, Primitives } from "../../../common/types/helpers";
import { generateBaseType } from "./baseTypeGen";

export function generateBase(
    input: Base,
    instance: ResourceInstance,
    helpers: SchemaHelpers
): Promise<Primitives> {
    if (Array.isArray(input)) {
        return Promise.all(
            (input as Base[]).map((entry) =>
                generateBase(entry, instance, helpers)
            )
        );
    }

    return generateBaseType(input, instance, helpers);
}
