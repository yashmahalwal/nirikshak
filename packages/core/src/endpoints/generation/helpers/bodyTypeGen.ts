import faker from "faker";
import {
    BodyInstance,
    BodyType,
    isBase,
    isWithModifiers,
    isOneOfEntries,
} from "../../types/helpers";
import { ResourceInstance } from "../../../resource/types/helper";
import { SchemaHelpers } from "../../../common/types/helpers";
import { generateBase } from "./baseGen";
import { generateWithModifiers } from "./withModifiersGen";
import { generateOneOfEntries } from "./oneOfGen";
import { isResourceString } from "../../types/resourceString";
import { getFromResource } from "../resourceStringGen";

export async function generateBodyType(
    input: BodyType,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<BodyInstance> {
    const o: BodyInstance = {};

    for (const key in input) {
        const entry = input[key];
        if (isResourceString(entry)) {
            try {
                o[key] = getFromResource(entry, resource);
            } catch {
                continue;
            }
        } else if (isBase(entry))
            o[key] = await generateBase(entry, resource, helpers);
        else if (isWithModifiers(entry)) {
            if (entry.optional && faker.random.boolean()) continue;

            o[key] = await generateWithModifiers(entry, resource, helpers);
        } else if (isOneOfEntries(entry)) {
            const selectedValue = generateOneOfEntries(entry);

            if (isWithModifiers(selectedValue.selection)) {
                if (selectedValue.selection.optional && faker.random.boolean())
                    continue;
                o[key] = await generateWithModifiers(
                    selectedValue.selection,
                    resource,
                    helpers
                );
            } else if (isBase(selectedValue.selection))
                o[key] = await generateBase(
                    selectedValue.selection,
                    resource,
                    helpers
                );
            else
                o[key] = await generateBodyType(
                    selectedValue.selection,
                    resource,
                    helpers
                );
        } else {
            o[key] = await generateBodyType(entry, resource, helpers);
        }
    }

    return o;
}
