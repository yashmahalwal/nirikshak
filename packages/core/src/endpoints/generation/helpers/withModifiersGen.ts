import {
    WithModifiers,
    BodyType,
    Base,
    isWithModifiersBase,
    BodyInstance,
} from "../../types/helpers";
import { ResourceInstance } from "../../../resource/types/helper";
import { SchemaHelpers, Primitives } from "../../../common/types/helpers";
import faker from "faker";
import { MIN_PLURAL_ENTRIES, MAX_PLURAL_ENTRIES } from "../../../common/Env";
import { generateBase } from "./baseGen";
import { generateBodyType } from "./bodyTypeGen";

export async function generateWithModifiersBase(
    input: WithModifiers<Base>,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<Primitives> {
    if (input.nullable && faker.random.boolean()) return null;

    if (input.plural) {
        const arr: Promise<Primitives>[] = [];

        const length = faker.random.number({
            min: MIN_PLURAL_ENTRIES,
            max: MAX_PLURAL_ENTRIES,
            precision: 1,
        });

        for (let i = 0; i < length; i++)
            arr.push(generateBase(input.type, resource, helpers));

        return Promise.all(arr);
    }

    return generateBase(input.type, resource, helpers);
}

export async function generateWithModifiersBodyType(
    input: WithModifiers<BodyType>,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<BodyInstance | BodyInstance[] | null> {
    if (input.nullable && faker.random.boolean()) return null;

    if (input.plural) {
        const arr: Promise<BodyInstance>[] = [];

        const length = faker.random.number({
            min: MIN_PLURAL_ENTRIES,
            max: MAX_PLURAL_ENTRIES,
            precision: 1,
        });

        for (let i = 0; i < length; i++)
            arr.push(generateBodyType(input.field, resource, helpers));

        return Promise.all(arr);
    }

    return generateBodyType(input.field, resource, helpers);
}

export async function generateWithModifiers<T extends BodyType | Base>(
    input: WithModifiers<T>,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<
    null | (T extends Base ? Primitives : BodyInstance[] | BodyInstance)
> {
    if (isWithModifiersBase(input))
        return generateWithModifiersBase(input, resource, helpers) as any;
    return generateWithModifiersBodyType(
        input as WithModifiers<BodyType>,
        resource,
        helpers
    ) as any;
}
