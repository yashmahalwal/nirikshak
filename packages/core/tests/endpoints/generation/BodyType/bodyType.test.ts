import faker from "faker";
import {
    BodyType,
    BodyInstance,
    Base,
    isWithModifiers,
} from "../../../../src/endpoints/types/helpers";
import { RANDOMNESS_ITERATIONS } from "../../../../src/common/Env";
import { generateBodyType } from "../../../../src/endpoints/generation/helpers/bodyTypeGen";
import { SchemaHelpers } from "../../../../src/common/types/helpers";
import { generateWithModifiers } from "../../../../src/endpoints/generation/helpers/withModifiersGen";
import { generateBase } from "../../../../src/endpoints/generation/helpers/baseGen";
import { generateOneOfEntries } from "../../../../src/endpoints/generation/helpers/oneOfGen";

const resource = {
    id: faker.random.uuid(),
    name: {
        firstName: faker.name.firstName(),
        lastName: faker.name.lastName(),
    },
    addresses: [
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
    ],
    recursion: {
        bases: {
            args: {
                plethora: [1, 2, 10, [[7, 5, [4]]], 3],
            },
        },
    },
};

const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidEntries: {
    input: BodyType;
    output: () => Promise<BodyInstance>;
}[] = [
    {
        input: {
            address: "resource:address",
        },
        output: async (): Promise<BodyInstance> => ({}),
    },
    {
        input: {
            sampleFirst: [
                1,
                2,
                3,
                ["faker:random.number", [[["custom:word"]]]],
            ],
            efforts: {
                essentials: {
                    properly: true,
                    improperly: false,
                    versions: [
                        "ES6",
                        "faker:system.semver",
                        "faker:system.semver",
                        "faker:system.semver",
                        {
                            function: "faker:lorem.lines",
                            args: [2],
                        },
                    ],
                },
            },
        },
        async output(): Promise<BodyInstance> {
            return {
                sampleFirst: [
                    1,
                    2,
                    3,
                    [faker.random.number(), [[[await Helpers.word()]]]],
                ],
                efforts: {
                    essentials: {
                        properly: true,
                        improperly: false,
                        versions: [
                            "ES6",
                            faker.system.semver(),
                            faker.system.semver(),
                            faker.system.semver(),
                            faker.lorem.lines(2),
                        ],
                    },
                },
            };
        },
    },
    {
        input: {
            id: "faker:random.word",
            places: {
                type: "custom:word",
                plural: true,
                nullable: true,
                optional: true,
            },
            age: {
                type: { function: "faker:random.number", args: [100] },
            },
            age1: {
                type: { function: "faker:random.number", args: [100] },
                nullable: true,
            },
            age2: {
                type: { function: "faker:random.number", args: [100] },
                optional: true,
            },
            age3: {
                type: { function: "faker:random.number", args: [100] },
                plural: true,
            },
        },
        async output(): Promise<BodyInstance> {
            const o: BodyInstance = {};
            for (const key in this.input) {
                switch (key) {
                    case "id":
                        o["id"] = faker.random.word();
                        break;
                    case "places":
                        // optional
                        if (faker.random.boolean()) break;
                        o["places"] = await generateWithModifiers(
                            this.input.places,
                            resource,
                            Helpers
                        );
                        break;
                    case "age":
                        o["age"] = faker.random.number(100);
                        break;
                    case "age1":
                        o["age1"] = await generateWithModifiers(
                            this.input.age1,
                            resource,
                            Helpers
                        );
                        break;
                    case "age2":
                        if (faker.random.boolean()) continue;
                        o["age2"] = await generateWithModifiers(
                            this.input.age2,
                            resource,
                            Helpers
                        );
                        break;
                    case "age3":
                        o["age3"] = await generateWithModifiers(
                            this.input.age3,
                            resource,
                            Helpers
                        );
                }
            }

            return o;
        },
    },
    {
        input: {
            id: "faker:random.word",
            addresses: "resource:addresses",
            places: {
                field: { value: { type: "custom:word" } },
                plural: true,
                nullable: true,
                optional: true,
            },
            age: {
                field: {
                    value: {
                        type: { function: "faker:random.number", args: [100] },
                    },
                },
            },
            age1: {
                field: { typeName: "faker:random.number" },
                nullable: true,
            },
            age2: {
                field: {
                    value: { function: "faker:random.number", args: [100] },
                },
                optional: true,
            },
            age3: {
                field: {
                    value: {
                        type: { function: "faker:random.number", args: [100] },
                    },
                },
                plural: true,
            },
        },
        async output(): Promise<BodyInstance> {
            const o: BodyInstance = {};
            for (const key in this.input) {
                switch (key) {
                    case "addresses":
                        o["addresses"] = resource.addresses;
                        break;
                    case "id":
                        o["id"] = faker.random.word();
                        break;
                    case "places":
                        // optional
                        if (faker.random.boolean()) break;
                        o["places"] = await generateWithModifiers(
                            this.input.places,
                            resource,
                            Helpers
                        );
                        break;
                    case "age":
                        o["age"] = await generateWithModifiers(
                            this.input.age,
                            resource,
                            Helpers
                        );
                        break;
                    case "age1":
                        o["age1"] = await generateWithModifiers(
                            this.input.age1,
                            resource,
                            Helpers
                        );
                        break;
                    case "age2":
                        if (faker.random.boolean()) continue;
                        o["age2"] = await generateWithModifiers(
                            this.input.age2,
                            resource,
                            Helpers
                        );
                        break;
                    case "age3":
                        o["age3"] = await generateWithModifiers(
                            this.input.age3,
                            resource,
                            Helpers
                        );
                }
            }

            return o;
        },
    },
    {
        input: {
            criteria: {
                fields: [
                    {
                        f1: {
                            types: ["faker:random.number", 1, 2, 3, 4, 5],
                        },
                        f2: {
                            type: "custom:word",
                            plural: true,
                        },
                    },
                    {
                        age: "custom:word",
                        address: "resource:addresses[0].city",
                    },
                ],
            },
            number: {
                types: [1, 2, 3, 4, 5, "resource:id", 7],
            },
            mixedEntries: {
                fields: [
                    {
                        field: { age: 10 },
                        optional: true,
                        nullable: true,
                        plural: true,
                    },
                    {
                        field: { name: "Purpose" },
                    },
                    {
                        class: "CSE",
                    },
                ],
                types: [
                    [1, 2, 3],
                    [4, 5, 6],
                    ["custom:word", "resource:id"],
                    { type: 17, nullable: true, plural: true },
                ],
            },
        },
        async output(): Promise<BodyInstance> {
            const o: BodyInstance = {};
            for (const key in this.input) {
                switch (key) {
                    case "criteria": {
                        const entry = generateOneOfEntries(this.input.criteria);
                        o[key] = await generateBodyType(
                            entry.selection as BodyType,
                            resource,
                            Helpers
                        );
                        break;
                    }
                    case "number": {
                        const entry = generateOneOfEntries(this.input.number);
                        o[key] = await generateBase(
                            entry.selection as Base,
                            resource,
                            Helpers
                        );
                        break;
                    }
                    case "mixedEntries":
                        {
                            const entry = generateOneOfEntries(
                                this.input.mixedEntries
                            );
                            if (isWithModifiers(entry.selection)) {
                                if (
                                    entry.selection.optional &&
                                    faker.random.boolean()
                                )
                                    continue;
                                o["mixedEntries"] = await generateWithModifiers(
                                    entry.selection,
                                    resource,
                                    Helpers
                                );
                            } else if (entry.type === "base")
                                o[key] = await generateBase(
                                    entry.selection,
                                    resource,
                                    Helpers
                                );
                            else
                                o[key] = await generateBodyType(
                                    entry.selection,
                                    resource,
                                    Helpers
                                );
                        }
                        break;
                }
            }
            return o;
        },
    },
];

const InvalidBodyTypes: any[] = [
    { entry: "custom:lorem.ipsum" },
    { entry: "custom:invalidHelper" },
];

describe("Body generation", () => {
    for (let i = 1; i <= RANDOMNESS_ITERATIONS; i++)
        ValidEntries.forEach((entry, index) =>
            test(`Valid body generation: ${index}, iteration: ${i}`, async () => {
                faker.seed(123 * i + index);
                const got = await generateBodyType(
                    entry.input,
                    resource,
                    Helpers
                );
                faker.seed(123 * i + index);
                const expected = await entry.output();

                expect(got).toEqual(expected);
            })
        );

    test.each(InvalidBodyTypes)(`Invalid body generation: %#`, (entry) =>
        expect(
            generateBodyType(entry, resource, Helpers)
        ).rejects.toMatchSnapshot()
    );
});
