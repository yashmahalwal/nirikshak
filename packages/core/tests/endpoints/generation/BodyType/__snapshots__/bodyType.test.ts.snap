// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body generation Invalid body generation: 0 1`] = `[Error: function: custom:lorem.ipsum, args: ]`;

exports[`Body generation Invalid body generation: 1 1`] = `[Error: Non literal type value from custom:invalidHelper]`;
